### DISCUSSION TITLE
Getting Started with OMSimulatorPython

### DISCUSSION THREAD

**Discussion by Markus2xj** [OpenModelica/OMSimulator#1304] [Q&A] [answered]
Hi! I am new to OMSimulator and ran into an error I have not been able to fix. I hope you can help me get started. I exported two basic models to fmu (from OMEdit) which I would like to co-simulate using Python. The code I used is based on the OpenModelicaUsersGuide, where I added the path to the exported fmu's. I got the following error:

> Failed to load DLL C:\Users\username\...\model-9qvtnr3d\temp\0002_fmu_test2\binaries\win64\fmu_test2.dll:
> The specified module could not be found.
> error:   [terminate] Model "model" is in wrong model state
> Failed to load functions for FMI 2.info:    fmi2Instantiate() failed
> Function "fmi2FreeInstance" is not provided by this FMU.
> Function "fmi2FreeInstance" is not provided by this FMU.

Thanks in advance.

code:
```
import OMSimulator as oms

oms.setTempDirectory('./temp/')
model = oms.newModel("model")
root = model.addSystem('root', oms.Types.System.SC)
fmu_path = "C:/Users/username/AppData/Local/Temp/OpenModelica/OMEdit/"

root.addSubModel('fmu_test1', fmu_path+'fmu_test1/fmu_test1.fmu')
root.addSubModel('fmu_test2', fmu_path+'fmu_test2/fmu_test2.fmu')

root.addConnection('fmu_test1.y', 'fmu_test2.u')
root.addConnection('fmu_test2.y', 'fmu_test1.u')

model.resultFile = 'results.mat'
model.stopTime = 5
model.fixedStepSize = 1e-4

model.instantiate()
```

**Comment by arun3688** [comment]
@Markus2xj this is a known issue and which version of OpenModelica you are using, I suggest you to download the latest nightly builds and then export your fmu, the problem should be resolved by itself

**Reply by Markus2xj** [reply]
Thanks for your prompt reply! I downloaded the latest one - This one I believe [OpenModelica-v1.24.0-dev-196-gd48837e5bc-64bit.exe](https://build.openmodelica.org/omc/builds/windows/nightly-builds/64bit/OpenModelica-v1.24.0-dev-196-gd48837e5bc-64bit.exe), exported the models again, and it raises the same error unfortunately. Do you know a version where you know it does work? Hopefully that resolves the issue.

**Reply by arun3688** [reply]
@Markus2xj can you share your model, so that i can export the fmu and see if i can reproduce the issue

**Reply by Markus2xj** [reply]
The models I used are quite basic, just as a poc for testing:

```
model fmu_test1
  Modelica.Blocks.Sources.Sine sine annotation(
    Placement(transformation(origin = {-30, 50}, extent = {{-10, -10}, {10, 10}})));
  Modelica.Blocks.Math.Add add annotation(
    Placement(transformation(origin = {2, 2}, extent = {{-10, -10}, {10, 10}}, rotation = 180)));
  Modelica.Blocks.Interfaces.RealOutput y annotation(
    Placement(transformation(origin = {100, 54}, extent = {{-10, -10}, {10, 10}}), iconTransformation(origin = {84, 52}, extent = {{-10, -10}, {10, 10}})));
  Modelica.Blocks.Interfaces.RealInput u annotation(
    Placement(transformation(origin = {106, 0}, extent = {{-20, -20}, {20, 20}}, rotation = 180), iconTransformation(origin = {92, 14}, extent = {{-20, -20}, {20, 20}})));
equation
  connect(add.u1, u) annotation(
    Line(points = {{14, -4}, {106, -4}, {106, 0}}, color = {0, 0, 127}));
  connect(sine.y, y) annotation(
    Line(points = {{-18, 50}, {100, 50}, {100, 54}}, color = {0, 0, 127}));
  connect(sine.y, add.u2) annotation(
    Line(points = {{-18, 50}, {22, 50}, {22, 8}, {14, 8}}, color = {0, 0, 127}));

annotation(
    uses(Modelica(version = "4.0.0")));
end fmu_test1;
```

```
model fmu_test2
  Modelica.Blocks.Interfaces.RealInput u annotation(
    Placement(transformation(origin = {106, 0}, extent = {{-20, -20}, {20, 20}}, rotation = 180), iconTransformation(origin = {106, 0}, extent = {{-20, -20}, {20, 20}}, rotation = 180)));
  Modelica.Blocks.Interfaces.RealOutput y annotation(
    Placement(transformation(origin = {100, 54}, extent = {{-10, -10}, {10, 10}}), iconTransformation(origin = {100, 54}, extent = {{-10, -10}, {10, 10}})));
  Modelica.Blocks.Math.Add add(k2 = -1)  annotation(
    Placement(transformation(origin = {2, 2}, extent = {{-10, -10}, {10, 10}}, rotation = 180)));
  Modelica.Blocks.Sources.Cosine cosine annotation(
    Placement(transformation(origin = {2, 46}, extent = {{-10, -10}, {10, 10}})));
equation
  connect(add.u1, u) annotation(
    Line(points = {{14, -4}, {106, -4}, {106, 0}}, color = {0, 0, 127}));
  connect(cosine.y, y) annotation(
    Line(points = {{14, 46}, {64, 46}, {64, 54}, {100, 54}}, color = {0, 0, 127}));
  connect(cosine.y, add.u2) annotation(
    Line(points = {{14, 46}, {24, 46}, {24, 8}, {14, 8}}, color = {0, 0, 127}));
  annotation(
    uses(Modelica(version = "4.0.0")));end fmu_test2;
```
Within OMEdit, the following model allowed co-simulation of the exported FMUs:
```
model fmu_test
  fmu_test1_me_FMU fmu_test1_me_FMU1 annotation(
    Placement(transformation(origin = {-62, 42}, extent = {{-10, -10}, {10, 10}})));
  fmu_test2_me_FMU fmu_test2_me_FMU1 annotation(
    Placement(transformation(origin = {-2, 42}, extent = {{-10, -10}, {10, 10}})));
equation
  connect(fmu_test1_me_FMU1.y, fmu_test2_me_FMU1.u) annotation(
    Line(points = {{-51, 49}, {-13, 49}}, color = {0, 0, 127}));
  connect(fmu_test2_me_FMU1.y, fmu_test1_me_FMU1.u) annotation(
    Line(points = {{9, 49}, {20, 49}, {20, 62}, {-90, 62}, {-90, 49}, {-73, 49}}, color = {0, 0, 127}));
end fmu_test;
```

**Comment by arun3688** [comment]
@Markus2xj I tested your example and it works for me, see the attached file with the fmu's and python script, What i see is you are trying to do co simulation, if you have exported your fmu with only co-simulation model then you should change the type of system to WC in your python script

`root = model.addSystem('root', oms.Types.System.WC) ## it should be WC and not SC`

I have exported the fmu as both model_exchange and co Simulation mode , so that should work. And also make sure you download the `OMSimulator` version `pip install OMSimulator`

[test.zip](https://github.com/user-attachments/files/16323651/test.zip)

I hope this should resolve the issue

**Reply by Markus2xj** [reply]
Thank you for coming back to me. I used your exported FMU's and it worked! Like you mentioned, it seems the problem originates from the fmu exports. For some reason my fmu_test1.fmu export does work together with the fmu_test2.fmu you provided, but the other way around raises the error. I am unfamiliar with exporting the fmu as a model_exchange or co Simulation mode, could you explain how you export your models (or provide a some reference materials)?  So far I have exported the models like the picture below (using the FMU option). Thanks again, your help is much appreciated.
![image](https://github.com/user-attachments/assets/4df28940-f09b-447d-b07a-8201df6aa992)

**Comment by arun3688** [comment]
@Markus2xj From OMEdit select `tools->Options->FMI ` and choose the fmu type

![image](https://github.com/user-attachments/assets/19cc9ae3-caff-4149-b76b-c67a858b5f7a)

**Reply by Markus2xj** [reply]
I tried the same settings, but unfortunately, also after restarting, the new fmus raise the same error in the script. It seems there is some difference in the fmu exports. What version of OMEdit are you using exactly? Another thing I noticed is that the size of my fmus are 630 KB, and yours are 658 KB for some reason.

**Comment by arun3688** [comment]
@Markus2xj my omc version is `v1.24.0-dev-183-g504611b14f-cmake` may be uninstall the older omc version and install a clean openmodelica version from nightly builds.

**Reply by Markus2xj** [reply]
Uninstalled, and reinstalled your version, unfortunately still the same error. I'm eager to learn what I am doing wrong.

![image](https://github.com/user-attachments/assets/3531debb-a523-4060-bae5-9d128496ecd4)

**Comment by arun3688** [comment]
@Markus2xj can you please upload your fmu's and i can see if i find something

**Reply by Markus2xj** [reply]
Attached here:
[fmu_tests.zip](https://github.com/user-attachments/files/16389435/fmu_tests.zip)
Thanks for looking into it!

**Comment by arun3688** [comment]
@Markus2xj I looked into your fmu's and i see one problem `libwinpthread-1.dll` is missing inside the `binaries` folder, can you just copy the missing dll from my fmu and put it inside your fmu and run it, i guess that should solve the problem, if that solves the problem check in the OpenModelica installation directory path `C:/OpenModelicaXXXX/bin` if `libwinpthread-1.dll` exists

**Reply by Markus2xj** [reply]
Hi @arun3688, adding the `libwinpthread-1.dll` from your file did the trick! Thank you. In the bin folder of OM, the `libwinpthread-1.dll` file does exist though. Let me know if there's something I can do to automatically add required files to the fmu. But at least you provided a workaround to produce fmu's.

**Reply by arun3688** [reply]
@Markus2xj this is really strange by default the libwinpthread-1.dll should be automatically exported to fmu's, i do not know what could be the reason @adrpo @casella can you please comment on this

**Reply by AnHeuermann** [reply, marked as answer]
I bet this is related to https://github.com/OpenModelica/OpenModelica/issues/12065. Maybe try `gcc` instead of `clang`.

**Reply by casella** [reply]
I'm afraid I really can't help with dll hell, sorry

**Reply by Markus2xj** [reply]
Thank you all for helping, changing to gcc and gcc++ did the trick to produce working fmus.
